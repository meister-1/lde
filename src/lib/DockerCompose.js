import yml from "yaml"
import fs from "fs-sync"

export default class DockerCompose {

    constructor(apps, outPath) {
        this.apps = apps
        this.outPath = outPath
        this.config = DockerCompose.getDefaultConfig()
    }

    static getDefaultConfig() {
        const dockerComposeConfig = {
            "version": "3.6",
            "services": {
                "nginx": {
                    "image": "nginx:latest",
                    "ports": [ "80:80", "443:443"]
                }
            },
            "networks": {
                "llnet": {
                    "driver": "overlay",
                    "ipam": {
                        "config": [
                            {
                                "subnet": "10.5.0.0/24"
                            }
                        ]
                    }
                }
            },
            "volumes": {}
        }

        return dockerComposeConfig
    }  

    createConfig() {

        const dockerComposeConfig = DockerCompose.getDefaultConfig()

        const dockerServerConfigs = this.apps.map(({domain, name, port, ip, version, volumes, attributes, variables}) => {
            
            const serverConfig = {
                "name": name,
                "domain": domain,
                "config": {
                    "image": `${name}:${version}`,
                    "ports": [`${port}:${port}`],
                    "networks": {
                        "llnet": {
                            "ipv4_address": `${ip}`
                        }
                    }
                }
            }
    
            if(volumes) {
    
                serverConfig["config"]["volumes"] = volumes
    
                volumes.map((el) => {
                    const source = el.split(":")[0]
    
                    if(source.indexOf("/") == -1) {
                        dockerComposeConfig["volumes"][source] = {}
                    }
                })
                
            }
            
            if(attributes) {
            
                Object.keys(attributes).forEach(key => {
                    serverConfig["config"][key] = attributes[key]
                })

            }
            
            if (variables) {    
    
                serverConfig["config"]["env_file"] = [`${name}/.env.tmp`]           
    
            }
    
            
            return serverConfig
    
        })
    
        dockerComposeConfig.services["nginx"]["depends_on"] = dockerServerConfigs.map(element => element.name)
        dockerComposeConfig.services["nginx"]["networks"] = {
            llnet: {
                aliases: dockerServerConfigs.map(element => element.domain)
            }
        }
        
        dockerServerConfigs.forEach((element) => {
            dockerComposeConfig.services[element.name] = element.config    
        })
        
        this.config = dockerComposeConfig

        return this
    }

    getConfig() {
        return yml.stringify(this.config)
    }

    writeFile() {
        if(this.config) {

            fs.write(`${this.outPath}/docker-compose.yml`, this.getConfig(), function(err) {
                if(err) {
                    console.log(err)
                }
            })

        }
    }

    static readConfigFile(outPath) {
        // console.log(`${outPath}/docker-compose.yml`)
        const dockerCompose = fs.read(`${outPath}/docker-compose.yml`)
        const dockerComposeConfig = yml.parse(dockerCompose)
        const apps = []
        
        for (const name in dockerComposeConfig.services) {
            
            const path = `${outPath}/${name}/.env`
            
            if(fs.exists(path)) {
                console.log(path)
            }
            
            apps.push({
                name,
                port: dockerComposeConfig.services[name].ports[0],
                image: dockerComposeConfig.services[name].image,
                envVarFilePath: fs.exists(path) ? path : null
            })
        }
    
        return apps
    
    }
}