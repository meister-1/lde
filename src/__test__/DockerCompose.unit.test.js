import DockerCompose from "../lib/DockerCompose"

const mockOneApp = [{
    ip: "10.5.0.10",
    domain: "a.meister1.com",
    port: "8000",
    sourceDir: "out/a/src",
    name: "a",
    version: "latest",
    gitUrl: "git@bitbucket.org:lokalleads/a.git",
    volumes: ["maven-data:/.m2"]
}]

const mockMultiApp = [{
  ip: "10.5.0.10",
  domain: "a.meister1.com",
  port: "8000",
  sourceDir: "out/a/src",
  name: "a",
  version: "latest",
  gitUrl: "git@bitbucket.org:lokalleads/a.git",
  volumes: ["maven-data:/.m2"]
},
{
  ip: "10.5.0.11",
  domain: "b.meister1.com",
  port: "8001",
  sourceDir: "out/b/src",
  name: "b",
  version: "latest",
  gitUrl: "git@bitbucket.org:lokalleads/b.git",
  volumes: ["maven-data:/.m2"],
  variables: `PORT=8001`
}]

describe('docker compose test', () => {
    test('create config yml with one app', () => {
      expect(new DockerCompose(mockOneApp).createConfig().getConfig()).toMatchSnapshot()
    })
    test('create config yml with multi apps', () => {
      expect(new DockerCompose(mockMultiApp).createConfig().getConfig()).toMatchSnapshot()
    })
})