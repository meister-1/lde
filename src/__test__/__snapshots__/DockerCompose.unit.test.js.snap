// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docker compose test create config yml with multi apps 1`] = `
"version: \\"3.6\\"
services:
  nginx:
    image: nginx:latest
    ports:
      - 80:80
      - 443:443
    depends_on:
      - a
      - b
    networks:
      llnet:
        aliases:
          - a.meister1.com
          - b.meister1.com
  a:
    image: a:latest
    ports:
      - 8000:8000
    networks:
      llnet:
        ipv4_address: 10.5.0.10
    volumes:
      - maven-data:/.m2
  b:
    image: b:latest
    ports:
      - 8001:8001
    networks:
      llnet:
        ipv4_address: 10.5.0.11
    volumes:
      - maven-data:/.m2
    env_file:
      - b/.env.tmp
networks:
  llnet:
    driver: overlay
    ipam:
      config:
        - subnet: 10.5.0.0/24
volumes:
  maven-data:
    {}
"
`;

exports[`docker compose test create config yml with one app 1`] = `
"version: \\"3.6\\"
services:
  nginx:
    image: nginx:latest
    ports:
      - 80:80
      - 443:443
    depends_on:
      - a
    networks:
      llnet:
        aliases:
          - a.meister1.com
  a:
    image: a:latest
    ports:
      - 8000:8000
    networks:
      llnet:
        ipv4_address: 10.5.0.10
    volumes:
      - maven-data:/.m2
networks:
  llnet:
    driver: overlay
    ipam:
      config:
        - subnet: 10.5.0.0/24
volumes:
  maven-data:
    {}
"
`;
